diff --git a/squashfs-tools/unsquash-2.c b/squashfs-tools/unsquash-2.c
--- a/squashfs-tools/unsquash-2.c
+++ b/squashfs-tools/unsquash-2.c
@@ -82,6 +82,11 @@ static int read_fragment_table(long long *table_start)
 		return FALSE;
 	}
 
+	if((long long) (size_t) bytes != bytes) {
+		ERROR("read_fragment_table: fragment table too large for address space\n");
+		return FALSE;
+	}
+
 	TRACE("read_fragment_table: %d fragments, reading %d fragment indexes "
 		"from 0x%llx\n", sBlk.s.fragments, indexes,
 		sBlk.s.fragment_table_start);
diff --git a/squashfs-tools/unsquash-3.c b/squashfs-tools/unsquash-3.c
--- a/squashfs-tools/unsquash-3.c
+++ b/squashfs-tools/unsquash-3.c
@@ -99,6 +99,11 @@ static int read_fragment_table(long long *table_start)
 		return FALSE;
 	}
 
+	if((long long) (size_t) bytes != bytes) {
+		ERROR("read_fragment_table: fragment table too large for address space\n");
+		return FALSE;
+	}
+
 	TRACE("read_fragment_table: %d fragments, reading %d fragment indexes "
 		"from 0x%llx\n", sBlk.s.fragments, indexes,
 		sBlk.s.fragment_table_start);
diff --git a/squashfs-tools/unsquash-4.c b/squashfs-tools/unsquash-4.c
--- a/squashfs-tools/unsquash-4.c
+++ b/squashfs-tools/unsquash-4.c
@@ -68,6 +68,11 @@ static int read_fragment_table(long long *table_start)
 		return FALSE;
 	}
 
+	if((long long) (size_t) bytes != bytes) {
+		ERROR("read_fragment_table: fragment table too large for address space\n");
+		return FALSE;
+	}
+
 	TRACE("read_fragment_table: %u fragments, reading %d fragment indexes "
 		"from 0x%llx\n", sBlk.s.fragments, indexes,
 		sBlk.s.fragment_table_start);
