diff --git a/squashfs-tools/unsquash-2.c b/squashfs-tools/unsquash-2.c
--- a/squashfs-tools/unsquash-2.c
+++ b/squashfs-tools/unsquash-2.c
@@ -154,7 +154,7 @@ static void read_fragment(unsigned int fragment, long long *start_block, int *s
 
 	squashfs_fragment_entry_2 *fragment_entry;
 
-	if(fragment > sBlk.s.fragments)
+	if(fragment >= sBlk.s.fragments)
 		EXIT_UNSQUASH("File system corrupted - fragment index in inode too large (fragment: %u)\n", fragment);
 
 	fragment_entry = &fragment_table[fragment];
diff --git a/squashfs-tools/unsquash-3.c b/squashfs-tools/unsquash-3.c
--- a/squashfs-tools/unsquash-3.c
+++ b/squashfs-tools/unsquash-3.c
@@ -164,7 +164,7 @@ static void read_fragment(unsigned int fragment, long long *start_block, int *s
 
 	squashfs_fragment_entry_3 *fragment_entry;
 
-	if(fragment > sBlk.s.fragments)
+	if(fragment >= sBlk.s.fragments)
 		EXIT_UNSQUASH("File system corrupted - fragment index in inode too large (fragment: %u)\n", fragment);
 
 	fragment_entry = &fragment_table[fragment];
