diff --git a/squashfs-tools/read_fs.c b/squashfs-tools/read_fs.c
--- a/squashfs-tools/read_fs.c
+++ b/squashfs-tools/read_fs.c
@@ -250,7 +250,7 @@ unsigned char *scan_inode_table(int fd, long long start, long long end,
 		case SQUASHFS_LREG_TYPE: {
 			struct squashfs_lreg_inode_header inode;
 			int frag_bytes, blocks, i;
-			long long start, file_bytes = 0;
+			long long start, file_bytes = 0, block_count;
 			unsigned int *block_list;
 
 			/*
@@ -269,9 +269,15 @@ unsigned char *scan_inode_table(int fd, long long start, long long end,
 			frag_bytes = inode.fragment == SQUASHFS_INVALID_FRAG ?
 				0 : inode.file_size % sBlk->block_size;
-			blocks = inode.fragment == SQUASHFS_INVALID_FRAG ?
-				(inode.file_size + sBlk->block_size - 1) >>
-				sBlk->block_log : inode.file_size >>
-				sBlk->block_log;
+			block_count = inode.file_size >> sBlk->block_log;
+			if(inode.fragment == SQUASHFS_INVALID_FRAG &&
+					inode.file_size % sBlk->block_size)
+				block_count++;
+			if(block_count > INT_MAX / sizeof(unsigned int)) {
+				ERROR("File system corrupted - too many blocks in inode (blocks: %lld)\n",
+					block_count);
+				goto corrupted2;
+			}
+			blocks = block_count;
 			start = inode.start_block;
 
 			TRACE("scan_inode_table: extended regular "
